Added step withSpanAttributes to set the attributes also on child spans, added setSpanAttributes for setting the attribute only on the target

christophe-kamphaus-jemmic opened the pull request:
The `withSpanAttributes` step was added. This new step sets the attributes on all child spans created inside the body block. This is implemented using the StepContext.

`setSpanAttributes` was added to set attribute on the target without a body block similar to the existing `withSpanAttribute`.
`setSpanAttributes` is preferred over `withSpanAttribute` for consistency with other plugins (eg. `withCredentials`).

Fixes https://github.com/jenkinsci/opentelemetry-plugin/issues/173

The main difficulty in implementing a solution for #173 was how to efficiently
* get child nodes of the target
  * I did not find an easy API for this.
  * I would have to either walk the whole flow graph or keep the span hierarchy somehow in memory. The OpenTelemetry API does not expose the parent context ID that I could find, so that was also not feasible.
  * In the end I just used the few `otelTraceService.getSpan` to set the attributes of some child spans of the root span and documented the limitations of the current implementation.
* remember the hierarchy of withSpanAttribute calls, so that new spans can efficiently set the required attributes
  * I solved this by adding the invisible action `OpenTelemetryAttributesAction` to the Run tracking child attributes for later agent and phase spans. The `OpenTelemetryAttributesAction` is added to the SpanContext to track attributes to be set to new spans within that context.
  * I adapted `OpenTelemetryAttributesAction` to be serializable to fix the error `Failed to serialize OpenTelemetryAttributesAction`.

#### Limitations

Some child spans that were created before the execution of setSpanAttributes / withSpanAttributes might be missed.
For example closed spans or some open child spans.
Child spans created after the execution of withSpanAttributes will all have the attribute set correctly.

Declarative pipelines can only use a single `withSpanAttributes` step inside the `options` block.

#### Potential future work (not in this PR)

* A step to create a new span (https://github.com/jenkinsci/opentelemetry-plugin/issues/90)

### Testing done

- [x] I added JUnit tests to verify the desired behavior.
  - Added test verifying the behavior of `withSpanAttributes` and `setSpanAttributes`
  - Added tests checks overriding behavior, ie. multiple `withSpanAttributes` steps with the same key.
- [x] Tested locally using the docker-compose in https://github.com/jenkinsci/opentelemetry-plugin/tree/main/demos 
  - using `withSpanAttributes([spanAttribute(key: 'build.tool', value: 'gradle', target: 'PIPELINE_ROOT_SPAN')]) { ... }` and verified in Jaeger that all expected spans have this tag.

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

christophe-kamphaus-jemmic commented:
@cyrille-leclerc would you mind taking a look?

christophe-kamphaus-jemmic commented:
Should the new parameter `setOn` of the `withSpanAttribute` step be documented in a Markdown somewhere?

kuisathaverat commented:
>get child nodes of the target
I did not find an easy API for this.
I would have to either walk the whole flow graph or keep the span hierarchy somehow in memory. The OpenTelemetry API does not expose the parent context ID that I could find, so that was also not feasible.
In the end I just used the few otelTraceService.getSpan to set the attributes of some child spans of the root span and documented the limitations of the current implementation.

the `trace.id`  identifies all the members in a transaction, so they are already related.

>remember the hierarchy of withSpanAttribute calls, so that new spans can efficiently set the required attributes
I solved this by adding invisible actions AttributeSetterAction to the Run and FlowNode's in a way that we can easily retrieve using the getAncestors method
Is this a good way to do this?
I noticed that these AttributeSetterAction are serialized. I refactor the code a bit to fix the error Failed to serialize AttributeSetterAction#attributeKey". Should AttributeSetterAction be serialized?

Serialized and memory consumption are the possible issues

kuisathaverat commented on src/main/java/io/jenkins/plugins/opentelemetry/job/OtelTraceService.java:
```diff
@@ -218,27 +219,36 @@ public void purgeRun(@NonNull Run run) {
     public void putSpan(@NonNull Run run, @NonNull Span span, @NonNull FlowNode flowNode) {
         // FYI for agent allocation, we have 2 FlowNodeMonitoringAction to track the agent allocation duration
         flowNode.addAction(new FlowNodeMonitoringAction(span));
+        run.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
```
This could be a problem on large pipelines with hundreds of children. Also, siblings in a deep tree will iterate over and over again over the upper levels.

christophe-kamphaus-jemmic replied:
I changed this to use SpanContext instead.
Only for some spans (agent allocation, finalise phase) a single `OpenTelemetryAttributesAction` on the Run is used for getting the attributes.

kuisathaverat commented:
>Is this a good way to do this?

probably not; there are other steps like withCredentials that propagate something to lower levels in the pipeline like environment variables; they enclose the children into a closure to make it obvious that you are propagation something to your children, and I doubt they use actions to propagate the context.

https://www.jenkins.io/doc/pipeline/steps/credentials-binding/
https://github.com/jenkinsci/credentials-binding-plugin/blob/master/src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

christophe-kamphaus-jemmic commented:
> there are other steps like withCredentials that propagate something to lower levels in the pipeline like environment variables

I will take a look how they do it. Thanks for the links.

christophe-kamphaus-jemmic commented:
> > Is this a good way to do this?
> 
> probably not; there are other steps like withCredentials that propagate something to lower levels in the pipeline like environment variables; they enclose the children into a closure to make it obvious that you are propagation something to your children, and I doubt they use actions to propagate the context.
> 
> https://www.jenkins.io/doc/pipeline/steps/credentials-binding/ https://github.com/jenkinsci/credentials-binding-plugin/blob/master/src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

I changed the implementation to not use actions set on the FlowNodes for context propagation.
Instead the `StepContext` now is used. This is also how [BindingStep](https://github.com/jenkinsci/credentials-binding-plugin/blob/master/src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java) does it.

christophe-kamphaus-jemmic commented:
> Declarative pipelines can only use a single withSpanAttribute step inside the options block.

To fix this limitation and to improve backwards compatibility, I could leave the behavior of `withSpanAttribute` unchanged and add a `withSpanAttributes` step that takes a list of attributes to be set.

christophe-kamphaus-jemmic commented:
> Should the new behavior of the withSpanAttribute step and the setSpanAttribute step be documented in a Markdown document somewhere?

* To be updated at least in https://github.com/jenkinsci/opentelemetry-plugin/blob/main/docs/job-traces.md
* Search for withSpanAttribute in *.md files

christophe-kamphaus-jemmic commented:
> > Declarative pipelines can only use a single withSpanAttribute step inside the options block.
> 
> To fix this limitation and to improve backwards compatibility, I could leave the behavior of `withSpanAttribute` unchanged and add a `withSpanAttributes` step that takes a list of attributes to be set.
> 
> What do you think?

After some reflection I've implemented this change.

`withSpanAttribute` is now backwards compatible.
The new `withSpanAttributes` can be used to set several span attributes on child spans. This also fixes the limitation of declarative pipelines only allowing a single step of a given type in the options block.

For consistency with other plugins (eg. `withCredentials`) and to better distinguish the behavior to `withSpanAttributes` I've added `setSpanAttributes` for only setting the given attributes on the target.

Should `withSpanAttribute` be marked as deprecated with a recommendation to use `setSpanAttributes` instead?

cyrille-leclerc commented on src/main/java/io/jenkins/plugins/opentelemetry/job/OtelTraceService.java:
```diff
@@ -233,9 +238,16 @@ public void putSpan(@NonNull Run run, @NonNull Span span) {
         LOGGER.log(Level.FINEST, () -> "putRunPhaseSpan(" + run.getFullDisplayName() + "," + OtelUtils.toDebugString(span) + ")");
     }
 
+    public void putAgentSpan(@NonNull Run run, @NonNull Span span, @NonNull FlowNode flowNode) {
```
Can we explain in the javadoc why we need a special handling for "agent allocation spans"

christophe-kamphaus-jemmic replied:
To keep it consistent with the existing `putSpan` methods, I've added the explanation as a regular comment instead of a javadoc. Is that fine with you @cyrille-leclerc ?

christophe-kamphaus-jemmic replied:
I also double-checked that `setAttributesToSpan(span, run.getAction(OpenTelemetryAttributesAction.class));` is indeed necessary for phase and agent spans. With that removed, some of the new tests in `WithSpanAttributesStepTest` would fail.

cyrille-leclerc commented:
Thank you for the effort, it took a long time. I suggested to add an explanation through javadocs.

kuisathaverat commented:
The code LGTM, I have to find some time to make a manual test to verify there is nothing weird.

kuisathaverat commented:
I'm sorry. It is still on my plate, but I do not find time to test it. I will try to find some gap this week.
